src/irmin-pack/s.ml

```diff
@@ -74,6 +74,62 @@ module type S = sig
 
   val stats :
     dump_blob_paths_to:string option -> commit:commit -> repo -> unit Lwt.t
+
+  module Snapshot : sig
+    type kinded_hash = Contents of hash * metadata | Node of hash
+    [@@deriving irmin]
+
+    type entry = { name : string; hash : kinded_hash } [@@deriving irmin]
+
+    type inode_tree = {
+      depth : int;
+      length : int;
+      pointers : (int * hash) list;
+    }
+    [@@deriving irmin]
+
+    type v = Inode_tree of inode_tree | Inode_value of entry list
+    [@@deriving irmin]
+
+    type inode = { v : v; root : bool } [@@deriving irmin]
+
+    type t = [ `Inode of inode | `Blob of Backend.Contents.Val.t ]
+    [@@deriving irmin]
+
+    type on_disk = [ `Reuse | `Path of string ]
+    (** The type for specifing an [on_disk] option when traversing a tree. The
+        traversal requires an index to keep track of visited elements.
+
+        - if [on_disk] is not specified, the index is in memory.
+        - if [on_disk] is [`Path path], the index is temporarily store at
+          [path].
+        - if [on_disk] is [`Reuse] the store's index is reused. *)
+
+    val export :
+      ?on_disk:on_disk ->
+      repo ->
+      (t -> unit Lwt.t) ->
+      root_key:Tree.kinded_key ->
+      int Lwt.t
+    (** [export ?on_disk repo f ~root_key] applies [f] to all inodes and
+        contents in a rooted tree, with root specified by [root_key]. All
+        elements are visited only once, in an unspecified but stable order.
```

Comment:
WDYT? I did not check in the code that what I describe is true.
```
The traversal order is stable. In [Inode_tree], it is lexicographic on the [index] function (see {TODO ref to irmin-pack-config}). 
In [Inode_value], it is lexicographic on the steps.

[f] is called in post-order. In other words, [f] is first called on the leaves, and the last 
call to [f] is on the root designated by [root_key].

The traversal skips objects that are structurally equal to objects that were already 
traversed. In other words, [export] internally uses a hash set in order to guarantee 
that all the objects passed to [f] don't hash the same way.
```

Reply:
thanks, its better